File: services/blog/nuldoc-src/slide/slide.ts
import { SlideError } from "../errors.ts";
import { Revision, stringToDate } from "../revision.ts";
import { z } from "zod/mod.ts";

export const SlideMetadataSchema = z.object({
  slide: z.object({
    uuid: z.string(),
    title: z.string(),
    event: z.string(),
    talkType: z.string(),
    link: z.string(),
    tags: z.array(z.string()),
    revisions: z.array(z.object({
      date: z.string(),
      remark: z.string(),
      isInternal: z.boolean().optional(),
    })),
  }),
});

export type SlideMetadata = z.infer<typeof SlideMetadataSchema>;

export type Slide = {
  sourceFilePath: string;
  uuid: string;
  title: string;
  event: string;
  talkType: string;
  slideLink: string;
  tags: string[];
  revisions: Revision[];
};

export function createNewSlideFromMetadata(
  { slide }: SlideMetadata,
  sourceFilePath: string,
): Slide {
  const revisions = slide.revisions.map(
    (rev, i) => {
      const date = rev.date;
      const remark = rev.remark;
      const isInternal = rev.isInternal ?? false;
      return {
        number: i + 1,
        date: stringToDate(date),
        remark,
        isInternal,
      };
    },
  );
  if (revisions.length === 0) {
    throw new SlideError(
      `[slide.new] 'slide.revisions' field is empty`,
    );
  }

  return {
    sourceFilePath: sourceFilePath,
    uuid: slide.uuid,
    title: slide.title,
    event: slide.event,
    talkType: slide.talkType,
    slideLink: slide.link,
    tags: slide.tags,
    revisions: revisions,
  };
}

File: services/nuldoc/lib/nuldoc/slide/slide.rb
module Nuldoc
  Slide = Data.define(
    :source_file_path,
    :uuid,
    :title,
    :event,
    :talk_type,
    :slide_link,
    :tags,
    :revisions
  )

  class SlideFactory
    def initialize(metadata, source_file_path)
      @metadata = metadata
      @source_file_path = source_file_path
    end

    def create
      slide_data = @metadata['slide']
      revisions = slide_data['revisions'].each_with_index.map do |rev, i|
        Revision.new(
          number: i + 1,
          date: Revision.string_to_date(rev['date']),
          remark: rev['remark'],
          is_internal: !rev['isInternal'].nil?
        )
      end

      raise "[slide.new] 'slide.revisions' field is empty" if revisions.empty?

      Slide.new(
        source_file_path: @source_file_path,
        uuid: slide_data['uuid'],
        title: slide_data['title'],
        event: slide_data['event'],
        talk_type: slide_data['talkType'],
        slide_link: slide_data['link'],
        tags: slide_data['tags'],
        revisions: revisions
      )
    end
  end
end
